### PR TITLE
Add ivar variants

### PR DESCRIPTION
**Pull request by andersgs**

### CODE REVIEW

**Comment by andersgs** [comment]
Weird it worked locally... I'll look in to it.

**Comment by andersgs** [comment]
For logging purposes, it looks like this was the issue:

```
N E X T F L O W  ~  version 20.11.0-edge
Launching `./tests/software/ivar/variants/main.nf` [desperate_meninsky] - revision: eca6b56c86
Creating Conda env: bioconda::ivar=1.3.1 [cache /home/runner/pytest_workflow_qfumrxbk/ivar_variants_no_gff/work/conda/env-b6deb684c62494fb6921b1962c660b51]
Error executing process > 'test_ivar_variants_no_gff:IVAR_VARIANTS (test)'

Caused by:
  Failed to create Conda environment
  command: conda create --mkdir --yes --quiet --prefix /home/runner/pytest_workflow_qfumrxbk/ivar_variants_no_gff/work/conda/env-b6deb684c62494fb6921b1962c660b51 bioconda::ivar=1.3.1
  status : 1
  message:
    InvalidArchiveError("Error with archive /home/runner/conda_pkgs_dir/libnghttp2-1.43.0-h812cca2_0.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nFailed to create dir 'lib'")
```

**Comment by andersgs** [comment]
> For logging purposes, it looks like this was the issue:
> 
> ```
> N E X T F L O W  ~  version 20.11.0-edge
> Launching `./tests/software/ivar/variants/main.nf` [desperate_meninsky] - revision: eca6b56c86
> Creating Conda env: bioconda::ivar=1.3.1 [cache /home/runner/pytest_workflow_qfumrxbk/ivar_variants_no_gff/work/conda/env-b6deb684c62494fb6921b1962c660b51]
> Error executing process > 'test_ivar_variants_no_gff:IVAR_VARIANTS (test)'
> 
> Caused by:
>   Failed to create Conda environment
>   command: conda create --mkdir --yes --quiet --prefix /home/runner/pytest_workflow_qfumrxbk/ivar_variants_no_gff/work/conda/env-b6deb684c62494fb6921b1962c660b51 bioconda::ivar=1.3.1
>   status : 1
>   message:
>     InvalidArchiveError("Error with archive /home/runner/conda_pkgs_dir/libnghttp2-1.43.0-h812cca2_0.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nFailed to create dir 'lib'")
> ```

This seem to fix itself by re-running... The docker and singularity images had incorrect URL and those are now fixed.

File: software/ivar/variants/main.nf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+process IVAR_VARIANTS {
+    tag "$meta.id"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
+
+    conda (params.enable_conda ? "bioconda::ivar=1.3.1" : null)
+
```
Comment (drpatelh) [inline software/ivar/variants/main.nf:15]:
```suggestion
```

File: software/ivar/variants/main.nf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    samtools mpileup \\
+        $options.args2 \\
+        --reference $fasta \\
+        $bam | \\
+        $save_mpileup  \\
+        ivar variants \\
+        $options.args \\
+        $gff \\
+        -r $fasta \\
+        -p $prefix
```
Comment (drpatelh) [inline software/ivar/variants/main.nf:46]:
```suggestion
            $options.args \\
            $features \\
            -r $fasta \\
            -p $prefix
```

File: software/ivar/variants/meta.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - bam:
+        type: file
+        description: A sorted (with samtools sort) and trimmed (with iVar trim) bam file
+        pattern: "*.bam"
+    - fasta:
+        type: file
+        description: The reference sequence used for mapping and generating the BAM file
+        pattern: "*.fa"
```
Comment (drpatelh) [inline software/ivar/variants/meta.yml:47]: Add entry for `gff` here.

File: software/ivar/variants/meta.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - variants:
```
Comment (drpatelh) [inline software/ivar/variants/meta.yml:54]:
```suggestion
    - tsv:
```

File: tests/software/ivar/variants/main.nf [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+params.save_mpileup = true
+
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:6]:
```suggestion
params.gff = false

```

File: tests/software/ivar/variants/main.nf [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+workflow test_ivar_variants_no_gff {
+    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+                file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref )
+}
+
+params.gff = file("${launchDir}/tests/data/gff/sarscov2/MN908947.3.gff3", checkIfExists: true)
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:17]:
```suggestion
params.gff = "${launchDir}/tests/data/gff/sarscov2/MN908947.3.gff3"
```

File: tests/software/ivar/variants/main.nf [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+workflow test_ivar_variants_with_gff {
+    
+    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+                file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref )
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:25]:
```suggestion
    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
    def gff = file(params.gff, checkIfExists: true)
    def input = []
    input = [ [ id:'test'], 
              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
    IVAR_VARIANTS ( input, ref, gff )
```

File: tests/software/ivar/variants/main.nf [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+params.save_mpileup = true
+
+include { IVAR_VARIANTS } from '../../../../software/ivar/variants/main.nf' addParams([:])
+
+workflow test_ivar_variants_no_gff {
+    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+                file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref )
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:14]:
```suggestion
    def ref   = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
    def dummy = file("${launchDir}/tests/data/dummy/dummy_file.txt", checkIfExists: true)
    def input = []
    input = [ [ id:'test'], 
              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
    IVAR_VARIANTS ( input, ref, dummy )
```

File: tests/software/ivar/variants/test.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+- name: ivar variants no gff
+  command: nextflow run ./tests/software/ivar/variants -entry test_ivar_variants_no_gff -c tests/config/nextflow.config
```
Comment (drpatelh) [inline tests/software/ivar/variants/test.yml:2]: Be good to add a test here with and without mpileup file output too.

**Review by drpatelh — APPROVED** [review]
Thanks @andersgs. Some niggly things with syntax, channel and variable declaration but otherwise looking great!

**Review by drpatelh — CHANGES_REQUESTED** [review]
Commenting to actually request changes :)

File: tests/software/ivar/variants/main.nf [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+workflow test_ivar_variants_no_gff_with_mpileup {
+    def ref   = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def dummy = file("${launchDir}/tests/data/dummy/dummy_file.txt", checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref, dummy )
+}
+
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:24]:
```suggestion

params.gff = "${launchDir}/tests/data/gff/sarscov2/MN908947.3.gff3"
```

File: tests/software/ivar/variants/main.nf [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+workflow test_ivar_variants_with_gff_with_mpileup {
+    
+    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def gff = file(params.gff, checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref, gff )
+}
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:33]:
```suggestion

params.gff = "${launchDir}/tests/data/gff/sarscov2/MN908947.3.gff3"

workflow test_ivar_variants_with_gff_with_mpileup {
    
    def ref = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
    def gff = file(params.gff, checkIfExists: true)
    def input = []
    input = [ [ id:'test'], 
              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
    IVAR_VARIANTS ( input, ref, gff )
}
```

File: tests/software/ivar/variants/main.nf [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { IVAR_VARIANTS } from '../../../../software/ivar/variants/main.nf' addParams([:])
+
+workflow test_ivar_variants_no_gff_no_mpileup {
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:7]:
```suggestion

params.gff = false

workflow test_ivar_variants_no_gff_no_mpileup {
```

File: tests/software/ivar/variants/main.nf [hunk 13]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { IVAR_VARIANTS } from '../../../../software/ivar/variants/main.nf' addParams([:])
+
+params.gff = false
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:7]:
```suggestion
params.gff          = false
params.save_mpileup = false
```

File: tests/software/ivar/variants/main.nf [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+workflow test_ivar_variants_no_gff_no_mpileup {
+    def ref   = file("${launchDir}/tests/data/fasta/sarscov2/MN908947.3.fa", checkIfExists: true)
+    def dummy = file("${launchDir}/tests/data/dummy/dummy_file.txt", checkIfExists: true)
+    def input = []
+    input = [ [ id:'test'], 
+              file("${launchDir}/tests/data/bam/test-sc2-artic-v3-sorted-trimmed.bam", checkIfExists: true) ]
+    IVAR_VARIANTS ( input, ref, dummy )
+}
+
+workflow test_ivar_variants_no_gff_with_mpileup {
```
Comment (drpatelh) [inline tests/software/ivar/variants/main.nf:19]:
```suggestion

params.gff          = false
params.save_mpileup = true

workflow test_ivar_variants_no_gff_with_mpileup {
```

**Comment by drpatelh** [comment]
Hmmm...something 🐟 going on here. Will merge and fix in a follow up PR. Thanks @andersgs  😎

**Review by drpatelh — APPROVED** [review]

**Pull request merged**